### PR TITLE
Core: Track data files by spec id instead of full PartitionSpec

### PR DESCRIPTION
**Pull request by nastra**

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 1]
```diff
@@ -138,20 +138,16 @@ protected boolean isCaseSensitive() {
...
+        newDataFilesBySpec.keySet().stream().map(this::spec).collect(ImmutableSet.toImmutableSet());
+    Preconditions.checkState(
+        !specs.isEmpty(), "Cannot determine partition specs: no data files have been added");
     Preconditions.checkState(
         specs.size() == 1,
         "Cannot return a single partition spec: data files with different partition specs have been added");
     return specs.iterator().next();
   }
 
-  protected Set<PartitionSpec> dataSpecs() {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:148]: this method hasn't been used anywhere else, so I think it's fine to inline it

**Review by singhpk234 — APPROVED** [review]
LGTM !

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 2]
```diff
@@ -138,20 +138,16 @@ protected boolean isCaseSensitive() {
   }
 
   protected PartitionSpec dataSpec() {
-    Set<PartitionSpec> specs = dataSpecs();
+    Set<PartitionSpec> specs =
+        newDataFilesBySpec.keySet().stream().map(this::spec).collect(ImmutableSet.toImmutableSet());
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:142]: This seems a bit over-complicated. At the end of this method, there will be just one spec, so why look up all the specs by ID and then fail? I think this should look more like this:

```java
  Set<Integer> specs = newDataFilesBySpec.keySet();
  Preconditions...
  return spec(Iterables.getOnlyElement(specs));
```

**Comment by rdblue** [comment]
One comment. Otherwise this look good.

**Review by amogh-jahagirdar — APPROVED** [review]

**Comment by amogh-jahagirdar** [comment]
The change looks good to me, I'll go ahead and merge since @rdblue comment was addressed. Thanks for the improvement @nastra , and for the reviews @singhpk234 @rdblue!

**Pull request merged**